src/db.rs

```diff
@@ -69,14 +73,43 @@ impl ChalkDatabase {
 
 impl ChalkSolveDatabase for ChalkDatabase {
     fn program_clauses_that_could_match(&self, goal: &DomainGoal, vec: &mut Vec<ProgramClause>) {
+        let mut clauses = vec![];
+        match goal {
+            DomainGoal::Holds(WhereClause::Implemented(trait_ref)) => {
+                self.trait_datum(trait_ref.trait_id)
+                    .to_program_clauses(self, &mut clauses);
+
+                // TODO sized, unsize_trait, builtin impls?
+            }
+            DomainGoal::Holds(WhereClause::ProjectionEq(projection_predicate)) => {
+                self.associated_ty_data(projection_predicate.projection.associated_ty_id)
+                    .to_program_clauses(self, &mut clauses);
+
+                // TODO filter out some clauses?
+            }
+            DomainGoal::WellFormed(WellFormed::Trait(trait_predicate)) => {
+                self.trait_datum(trait_predicate.trait_id)
+                    .to_program_clauses(self, &mut clauses);
+            }
+            DomainGoal::WellFormed(WellFormed::Ty(ty)) => match_ty(self, goal, ty, &mut clauses),
+            DomainGoal::FromEnv(_) => (), // Computed in the environment
+            DomainGoal::Normalize(projection_predicate) => {
+                self.associated_ty_data(projection_predicate.projection.associated_ty_id)
+                    .to_program_clauses(self, &mut clauses);
+            }
+            _ => (), // TODO rustc has just 4 enum variants, what about other Chalk DomainGoal variants?
+        };
+
         if let Ok(env) = self.environment() {
-            vec.extend(
-                env.program_clauses
-                    .iter()
-                    .filter(|&clause| clause.could_match(goal))
-                    .cloned(),
-            );
+            env.program_clauses_for_env(self, &mut clauses);
```

Comment:
Oops, easy mistake to make though, since they're identical structs modulo names :)

Good tip on inlining the logic, that also removed `ChalkSolveDatabase` in one go.